fml/unique_fd.h

```diff
@@ -22,10 +25,48 @@ namespace internal {
 
 namespace os_win {
 
+struct DirCacheEntry {
+  std::wstring filename;
+  FILE_ID_128 id;
+};
+
+// The order of these is important.  Must come before UniqueFDTraits struct
+// else linker error.  Embedding in struct also causes linker error.
+
 struct UniqueFDTraits {
+  inline static std::mutex file_map_mutex;
+  inline static std::map<HANDLE, DirCacheEntry> file_map;
+
   static HANDLE InvalidValue() { return INVALID_HANDLE_VALUE; }
   static bool IsValid(HANDLE value) { return value != InvalidValue(); }
-  static void Free(HANDLE fd);
+  static void Free_Handle(HANDLE fd);
+
+  static void Free(HANDLE fd) {
+    RemoveCacheEntry(fd);
+
+    UniqueFDTraits::Free_Handle(fd);
+  }
+
+  static void RemoveCacheEntry(HANDLE fd) {
+    const std::lock_guard<std::mutex> lock(file_map_mutex);
+
+    if (file_map.count(fd) > 0) {
+      file_map.erase(fd);
+    }
+  }
+
+  static void StoreCacheEntry(HANDLE fd, DirCacheEntry state) {
+    const std::lock_guard<std::mutex> lock(file_map_mutex);
+    file_map[fd] = state;
+  }
+
+  static std::optional<DirCacheEntry> GetCacheEntry(HANDLE fd) {
+    const std::lock_guard<std::mutex> lock(file_map_mutex);
+    if (file_map.count(fd) > 0) {
```

Comment:
The pattern we usually follow is:

```
auto found = file_map.find(fd);
return found == file_map.end() ? std::nullopt : *file_map;
```

Doesn't really change the complexity but we can just perform one lookup.

Reply:
Will fix in next commit.

Reply:
The *file_map part doesn't work hence:   return found == file_map.end() ? std::nullopt : std::optional<DirCacheEntry>{found->second};